### PR TITLE
Proof of concept workaround for connection freeze seen in JENKINS-60536

### PR DESCRIPTION
**Pull request by TaraWij**
Can you please review these changes from David France?

https://issues.jenkins-ci.org/browse/JENKINS-60536

Thank you very much in advance.

### CODE REVIEW

File: src/main/java/com/atlassian/httpclient/apache/httpcomponents/SettableFuturePromiseHttpPromiseAsyncClient.java [hunk 1]
```diff
@@ -39,24 +44,38 @@
             void doCompleted(final HttpResponse httpResponse)
             {
                 executor.execute( () -> future.set(httpResponse));
+                log.info( "Closing in doCompleted()" );
```
Comment (warden) [inline src/main/java/com/atlassian/httpclient/apache/httpcomponents/SettableFuturePromiseHttpPromiseAsyncClient.java]: `log.debug` probably? same for the two below (or remove them altogether)

File: src/main/java/com/atlassian/httpclient/apache/httpcomponents/ApacheAsyncHttpClient.java [hunk 2]
```diff
@@ -137,121 +137,128 @@ public ApacheAsyncHttpClient(final Supplier<String> applicationName,
...
+                .setConnectionManager(connectionManager)
+                .setRedirectStrategy(new RedirectStrategy())
+                .setUserAgent(getUserAgent(options))
+                .setDefaultRequestConfig(requestConfig);
+
+        if(Jenkins.getInstance() != null) {
+            ProxyConfiguration proxyConfiguration = Jenkins.getInstance().proxy;
+            if ( proxyConfiguration != null ) {
+                final HttpHost proxy = new HttpHost( proxyConfiguration.name, proxyConfiguration.port );
+                //clientBuilder.setProxy( proxy );
```
Comment (warden) [inline src/main/java/com/atlassian/httpclient/apache/httpcomponents/ApacheAsyncHttpClient.java]: I just noted this (which results of #154) - isn't it an unintentional change that actually disables proxy usage at all? 
maybe @escoem remembers that..

File: src/main/java/com/atlassian/httpclient/apache/httpcomponents/ApacheAsyncHttpClient.java [hunk 3]
```diff
@@ -469,6 +477,18 @@ private void publishEvent(Request request, long requestDuration, Throwable ex)
 
     private PromiseHttpAsyncClient getPromiseHttpAsyncClient(Request request)
     {
+
+        log.info( "Creating new HttpAsyncClient" );
```
Comment (warden) [inline src/main/java/com/atlassian/httpclient/apache/httpcomponents/ApacheAsyncHttpClient.java]: log.debug

File: src/main/java/com/atlassian/httpclient/apache/httpcomponents/SettableFuturePromiseHttpPromiseAsyncClient.java [hunk 4]
```diff
@@ -39,24 +44,38 @@
...
             }
         });
         return Promises.forListenableFuture(future);
     }
 
+    private void closeClient() {
+        try {
+            client.close();
+        } catch ( IOException e ) {
+            log.error( "Close failed" , e );
```
Comment (warden) [inline src/main/java/com/atlassian/httpclient/apache/httpcomponents/SettableFuturePromiseHttpPromiseAsyncClient.java]: How about making it a bit more meaningful, like `("Failed to close HttpAsyncClient", e);` ?

**Comment by artkoshelev** [comment]
@warden as far as i can see Connection Pool is staying in place, it's just pretty confusing git diff showing the same lines as deleted and added

**Comment by warden** [comment]
@artkoshelev do you think we should merge it as-is right now? (i can rebase and fix the commit message myself)

**Comment by artkoshelev** [comment]
I'd prefer to either see the normal diff or get comments about what's the actual change from the author first.

**Comment by olamy** [comment]
@warden definitely not ready to be merged....

**Pull request closed without merging**